### DISCUSSION TITLE
Potential incompatibility between composite actions and containerized jobs?

### DISCUSSION THREAD

**Discussion by jasonrudolph** [community/community#13718] [Actions] [answered]
**tl;dr** -- When publishing a composite action, is there a way to ensure that the action always runs on a [GitHub-hosted runner type](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#choosing-github-hosted-runners) so that we can be sure the action's dependencies are present at runtime?

---

I have a [composite action](https://docs.github.com/en/actions/creating-actions/about-custom-actions#composite-actions) that's been running great on hosted runners. Here's an intentionally simplified example:

## Composite action

```yaml
runs:
  using: "composite"
  steps:
    - run: curl http://example.com
      shell: bash
```

## Workflow using the composite action

```yaml
on: [push]

jobs:
  demo:
    runs-on: ubuntu-latest

    steps:
      - uses: some-user/some-action@main
```

This has been working nicely on all [GitHub-hosted runner types](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#choosing-github-hosted-runners) (e.g., `ubuntu-latest`, `windows-latest`, etc.).

For example, we can see a successful run of a simplified demo at https://github.com/jasonrudolph/composite-action-vs-containerized-job/runs/5725041450?check_suite_focus=true#step:3:3.

## Workflow using the composite action in containerized job

However, the example above breaks when you include the composite action in a [job that runs in a container](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idcontainer):

```yaml
on: [push]

jobs:
  demo:
    runs-on: ubuntu-latest
    container: ubuntu:latest

    steps:
      - uses: some-user/some-action@main
```

The action fails with the following error as seen at https://github.com/jasonrudolph/composite-action-vs-containerized-job/runs/5725052762?check_suite_focus=true#step:4:6:

```
/__w/_temp/369f0354-2e3e-4571-abf1-acc7e2c89e82.sh: line 1: curl: command not found
Error: Process completed with exit code 127.
```

## Composite actions seem to inherit the job's container

According to the [docs](https://docs.github.com/en/actions/using-workflows/workflow-syntax-for-github-actions#jobsjob_idcontainer), when a job specifies a container, that container will "run any steps in [the] job that don't already specify a container." Since the `ubuntu:latest` Docker image doesn't include `curl`, that explains why we get the error above.

## What should composite action publishers do in this situation?

When a team is defining their job, it makes sense that they can define a custom container to run their job's steps, and of course third-party action publishers can't know what utilities will be present in that container.

With that constraint in mind, what should composite action authors do in this situation. Specifically:

🙋🙋🙋 If I'm publishing a composite action that performs post-processing at the end of a job, how can I be sure that the action is running in an environment that satisfies the job's dependencies? Is there a way to have the composite action run on a GitHub-hosted runner type instead of inheriting the job's custom container?

**Comment by chrispat** [comment, marked as answer]
No that is not possible.  Your best bet is to check for and document your dependencies like `curl` in your action and provide a message when you fail.  Given customers can configure their own runners you can never guarantee that a hosted runner is even available.
